stdconv/string.hpp

```diff
@@ -0,0 +1,42 @@
+#pragma once
+
+#include "libmodmqttconv/converter.hpp"
+#include <boost/algorithm/string/predicate.hpp>
+
+class StringConverter : public DataConverter {
+    public:
+        virtual MqttValue toMqtt(const ModbusRegisters& data) const {
+            std::vector<uint16_t> registers = data.values();
+            ensureByteOrder(registers);
+            return MqttValue::fromBinary(registers.data(), data.getCount() * sizeof(std::uint16_t));
+        }
+
+        virtual ModbusRegisters toModbus(const MqttValue& value, int registerCount) const {
+            const uint16_t* source = static_cast<const uint16_t*>(value.getBinaryPtr());
+            const int elementCount = value.getBinarySize() / sizeof(std::uint16_t);
+            std::vector<uint16_t> registers(source, source + elementCount);
+            ensureByteOrder(registers);
+            registers.resize(std::min(elementCount, registerCount));
+            return ModbusRegisters(registers);
+        }
+
+        virtual void setArgs(const std::vector<std::string>& args) {
+            if (args.size() == 0) {
+                return;
+            }
+            std::string encoding = ConverterTools::getArg(0, args);
+            if (boost::algorithm::ends_with(encoding, "-be")) {
```

Comment:
It's implemented it as suggested, but I'm no longer convinced that this implementation is appropriate.

My concern is that the string converter never interprets or changes an _encoding_. It either takes the bytes as they are or swaps them. This is a endian conversion, not more and not less. The bytes are then handed over as binary to `MqttValue` which returns them as `std::string`. As far as I know, `std::string` is also encoding-agnostic, i.e. it contains no encoding information but raw bytes only.

Since no encoding interpretation or transformation is done at any point, the `ascii` in `ascii16-be` is wrong. I suggest to rename the configuration argument `encoding` to `byte-order` with allowed values `little-endian` (default, currently `none`) and `big-endian` (currently `ascii16-be`). What do you think?

Reply:
MqttValue returns std::string, which is either sent as mqtt payload or passed to RapidJson to construct mqtt payload. In both cases we want to send utf-8 encoded data by default.

In `none` case we assume that device already holds utf-8 compatible data and we just pass it through std.string coverter as mqtt payload.

In `ascii16-be` case we need to decode it first, by swapping pair of bytes. After that we still assume that this is ascii (hence utf-8 compatible) string. This is some kind of decoding.

Suppose that there is some modbus device that holds string data as utf16-le. It is one char per register. In this case to support it we will have to add icu library or use boost::locale to convert data to utf-8 before passing it to MqttValue.

`little-endian`, `big-endian` has some meaning if text characters are stored as multiple bytes, as in utf-16 example. Thats why I think that `ascii16-be` is not the best name, because it suggests that single ascii char is stored on two bytes. Unfortunately I do not have any better name for this encoding. 

I have no experience with modbus devices that store strings in registers, so maybe this encoding is used only in SBC devices and nowhere else? In this case I think `sbc-string` or something like that would be a better name.

Reply:
> `little-endian`, `big-endian` has some meaning if text characters are stored as multiple bytes, as in utf-16 example. Thats why I think that `ascii16-be` is not the best name, because it suggests that single ascii char is stored on two bytes. 

OK, that makes sense to me.

The problem is that the byte order setting does not depend on the datatype `string` (or this special implementation `sbc-string`) but on the device or maybe even the protocol implementation of a manufacturer. If you have a look into the manual, you'll see that not only the string (reg. 7-14) , but all multi-register values are "highest byte first" (baud rate, serial number & energy). That means this byte swapping logic needs to be applied to other datatypes, too:
* It's already part of the the `divide()` converter.
* Another popular energy meter, Eastron SDM120 ([manual](https://data.xn--stromzhler-v5a.eu/eastron/SDM120Modbus_manual.pdf)), offers many 4-byte float values that are provided "highest byte first". Suppose we add this string converter and later on a float converter. Then the float converter requires the same logic again.

It's rather some kind of filter, independent of the converter:

> Modbus device <-> Filter (byte swapping) <-> Converter (bytes ./. value) <-> MQTT

If we put this byte-swapping logic into each converter, it will be scattered across the code. Maybe inheritance could help (e.g. a base converter with a `low_first` parameter), but this would probably be incompatible to current converters. Or is it maybe possible to chain converters?